compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -1248,6 +1248,10 @@ impl<'rt, 'tcx, M: Machine<'tcx>> ValueVisitor<'tcx, M> for ValidityVisitor<'rt,
                     // Range patterns are precisely reflected into `valid_range` and thus
                     // handled fully by `visit_scalar` (called below).
                     ty::PatternKind::Range { .. } => {},
+
+                    // FIXME(pattern_types): check that the value is covered by one of the variants.
+                    // The layout may pessimistically cover actually illegal ranges.
+                    ty::PatternKind::Or(_patterns) => {}
```

Comment:
Is it clear how one would do this? Do we have to try one after the other? That seems very tricky as it means we have to validate the value "speculatively" where we ignore errors, until we know that all paths error, then we report one of them... as long as it's just integers this is fairly harmless, but something like `Ok(true) | Err(5) | Err(7)` should ideally be checked by "read discriminant, then proceed with the pattern restricted to what we see there" -- basically, compiling the pattern to a discrimination tree similar to what happens during match lowering, and executing that here.

Reply:
I think we'll have to do something similar to whatever we'll want to do if pattern types can have user defined trait impls. We'll need to detect overlap and possibly exhaustiveness some day. I leave this problem to future me, because it's not necessary for the MVP of replacing the rustc_scalar_layout attributes

Reply:
Checking the validity of `NonZeroI32` is definitely necessary, Miri must be able to detect that UB.

Checking the validity of general or patterns can be deferred, though I'd prefer if we did not use any or patterns in std that Miri can't support -- at least not for code that can be reached on stable.

Reply:
This is supported via the layout. That's what I was trying to convey with the comment.

Reply:
Okay so if the "or" pattern happens to fall into the fragment layout can handle, then Miri can check validity. If not, we don't actually use this information for codegen, so it's okay for Miri to not check validity. I take it patterns never reduce the size of the type, only the set of allowed values?

Reply:
Yes